python/pyspark/pandas/groupby.py

```diff
@@ -895,6 +895,89 @@ def sem(col: Column) -> Column:
             bool_to_numeric=True,
         )
 
+    # TODO: 1, 'n' accepts list and slice; 2, implement 'dropna' parameter
+    def nth(self, n: int) -> FrameLike:
+        """
+        Take the nth row from each group.
+
+        .. versionadded:: 3.4.0
+
+        Parameters
+        ----------
+        n : int
+            A single nth value for the row
+
+        Examples
+        --------
+
+        >>> df = ps.DataFrame({'A': [1, 1, 2, 1, 2],
+        ...                    'B': [np.nan, 2, 3, 4, 5]}, columns=['A', 'B'])
+        >>> g = df.groupby('A')
+        >>> g.nth(0)
+             B
+        A
+        1  NaN
+        2  3.0
+        >>> g.nth(1)
+             B
+        A
+        1  2.0
+        2  5.0
+        >>> g.nth(-1)
+             B
+        A
+        1  4.0
+        2  5.0
+
+        See Also
+        --------
+        pyspark.pandas.Series.groupby
+        pyspark.pandas.DataFrame.groupby
+        """
+        groupkey_names = [SPARK_INDEX_NAME_FORMAT(i) for i in range(len(self._groupkeys))]
+        internal, agg_columns, sdf = self._prepare_reduce(
+            groupkey_names=groupkey_names,
+            accepted_spark_types=None,
+            bool_to_numeric=False,
+        )
+        psdf: DataFrame = DataFrame(internal)
+
+        if len(psdf._internal.column_labels) > 0:
+            window1 = Window.partitionBy(*groupkey_names).orderBy(NATURAL_ORDER_COLUMN_NAME)
+            tmp_row_number_col = "__tmp_row_number_col__"
+            if n >= 0:
+                sdf = (
+                    psdf._internal.spark_frame.withColumn(
+                        tmp_row_number_col, F.row_number().over(window1)
+                    )
+                    .where(F.col(tmp_row_number_col) == n + 1)
+                    .drop(tmp_row_number_col)
+                )
+            else:
+                window2 = Window.partitionBy(*groupkey_names).rowsBetween(
+                    Window.unboundedPreceding, Window.unboundedFollowing
+                )
+                tmp_group_size_col = "__tmp_group_size_col__"
+                sdf = (
+                    psdf._internal.spark_frame.withColumn(
+                        tmp_group_size_col, F.count(F.lit(0)).over(window2)
+                    )
+                    .withColumn(tmp_row_number_col, F.row_number().over(window1))
+                    .where(F.col(tmp_row_number_col) == F.col(tmp_group_size_col) + 1 + n)
+                    .drop(tmp_group_size_col, tmp_row_number_col)
+                )
+        else:
+            sdf = sdf.select(*groupkey_names).distinct()
```

Comment:
Add a test to cover this? I'm a little fuzzy about this

Reply:
there seems a bug in Pandas' `GroupBy.nth`, its returned index varies with `n`:
```
In [23]: pdf
Out[23]: 
   A    B  C      D
0  1  3.1  a   True
1  2  4.1  b  False
2  1  4.1  b  False
3  2  3.1  a   True

In [24]: pdf.groupby(["A", "B", "C", "D"]).nth(0)
Out[24]: 
Empty DataFrame
Columns: []
Index: [(1, 3.1, a, True), (1, 4.1, b, False), (2, 3.1, a, True), (2, 4.1, b, False)]

In [25]: pdf.groupby(["A", "B", "C", "D"]).nth(0).index
Out[25]: 
MultiIndex([(1, 3.1, 'a',  True),
            (1, 4.1, 'b', False),
            (2, 3.1, 'a',  True),
            (2, 4.1, 'b', False)],
           names=['A', 'B', 'C', 'D'])

In [26]: pdf.groupby(["A", "B", "C", "D"]).nth(1)
Out[26]: 
Empty DataFrame
Columns: []
Index: []

In [27]: pdf.groupby(["A", "B", "C", "D"]).nth(1).index
Out[27]: MultiIndex([], names=['A', 'B', 'C', 'D'])

In [28]: pdf.groupby(["A", "B", "C", "D"]).nth(-1)
Out[28]: 
Empty DataFrame
Columns: []
Index: [(1, 3.1, a, True), (1, 4.1, b, False), (2, 3.1, a, True), (2, 4.1, b, False)]

In [29]: pdf.groupby(["A", "B", "C", "D"]).nth(-1).index
Out[29]: 
MultiIndex([(1, 3.1, 'a',  True),
            (1, 4.1, 'b', False),
            (2, 3.1, 'a',  True),
            (2, 4.1, 'b', False)],
           names=['A', 'B', 'C', 'D'])

In [30]: pdf.groupby(["A", "B", "C", "D"]).nth(-2)
Out[30]: 
Empty DataFrame
Columns: []
Index: []

In [31]: pdf.groupby(["A", "B", "C", "D"]).nth(-2).index
Out[31]: MultiIndex([], names=['A', 'B', 'C', 'D'])
```

while other functions' behavior in Pandas and PS are like this:
```
In [17]: pdf
Out[17]: 
   A    B  C      D
0  1  3.1  a   True
1  2  4.1  b  False
2  1  4.1  b  False
3  2  3.1  a   True

In [18]: pdf.groupby(["A", "B", "C", "D"]).max()
Out[18]: 
Empty DataFrame
Columns: []
Index: [(1, 3.1, a, True), (1, 4.1, b, False), (2, 3.1, a, True), (2, 4.1, b, False)]

In [19]: pdf.groupby(["A", "B", "C", "D"]).mad()
Out[19]: 
Empty DataFrame
Columns: []
Index: [(1, 3.1, a, True), (1, 4.1, b, False), (2, 3.1, a, True), (2, 4.1, b, False)]

In [20]: psdf.groupby(["A", "B", "C", "D"]).max()
Out[20]: 
Empty DataFrame
Columns: []
Index: [(1, 3.1, a, True), (2, 4.1, b, False), (1, 4.1, b, False), (2, 3.1, a, True)]

In [21]: psdf.groupby(["A", "B", "C", "D"]).mad()
Out[21]: 
Empty DataFrame
Columns: []
Index: [(1, 3.1, a, True), (2, 4.1, b, False), (1, 4.1, b, False), (2, 3.1, a, True)]

In [22]: 

In [22]: psdf.groupby(["A", "B", "C", "D"]).nth(0)
Out[22]: 
Empty DataFrame
Columns: []
Index: [(1, 3.1, a, True), (2, 4.1, b, False), (1, 4.1, b, False), (2, 3.1, a, True)]
```

Reply:
so I think we can not add a test for it for now

Reply:
If there is a bug in pandas, maybe we should add a test by manually creating the expected result rather than just skipping the test ?

e.g.
https://github.com/apache/spark/blob/6d2ce128058b439094cd1dd54253372af6977e79/python/pyspark/pandas/tests/test_series.py#L1654-L1658

Reply:
Oh... I just noticed that we're following the pandas behavior even though there is a bug in pandas.

When there is a bug in pandas, we usually do something like this:

- we don't follow the behavior of pandas, just we assume it works properly and implement it.
- comment the link related [pandas issues](https://github.com/pandas-dev/pandas/issues) to the test, from pandas repository(`https://github.com/pandas-dev/pandas/issues/...`) as below:
  - https://github.com/apache/spark/blob/0830575100c1bcbc89d98a6859fe3b8d46ca2e6e/python/pyspark/pandas/tests/data_type_ops/test_num_ops.py#L510-L517
  - https://github.com/apache/spark/blob/0830575100c1bcbc89d98a6859fe3b8d46ca2e6e/python/pyspark/pandas/tests/data_type_ops/test_num_ops.py#L478-L483

- If it's not clear that it's a bug (unless it's not an officially discussed as a bug in pandas community), we can just follow the pandas behavior.

Reply:
Or we can post a question for pandas community if it's a bug or intended behavior, and comment the question link it if they reply "yes, it's a bug".